### ISSUE TITLE
Supported precomputed component scanning 

### ISSUE THREAD

**Issue by graemerocher** [micronaut-projects/micronaut-core#911]
Suggestion is to add support for a `@ComponentScan` annotation defined on a bean (typically the `Application` bean) that is computed at compilation time, avoiding the need to do so at runtime and improving startup time.

Doing this would also avoid confusing behaviour currently around creating tests in subpackages.

Entity scanning is currently supported via `Environment.scan(..)` which executes at runtime this adds overhead if a user is using JPA or GORM entity scanning (or any other component scanning for that matter).

Using an annotation mapper and the existing ASM class path scanning we should be able to precompute `@Entity` scanning and any other scanning requirements. Only challenge is combining with with sources that are being compiled. For that we can use `RoundEnvironment.getElementsAnnotatedWith(..)` in APT and a manual search of the Groovy ModuleNode.

Note: Once implement the existing `scan(..)` method could possibly be deprecated.

**Comment by jameskleeh** [comment]
Perhaps the annotation should be supported on a package-info instead of a bean?

**Comment by graemerocher** [comment]
Yes we could make that work too

**Comment by graemerocher** [comment]
Resolved by #1276 

Can be achieved with:

```java
@Introspected(packages="foo.bar", includedAnnotations=Entity.class)
```

Then

```java
BeanIntrospector.findIntrospections(Entity.class)
```

**Comment by edalex-joe** [comment]
I'm using this incorrectly I think - does this do something similar to Spring's @ComponentScan? I've got a Micronaut HTTP @Client class in a library I import, but I haven't been able to have it picked up for injection into the main app.

I've tried using @Introspected(includedAnnotations = Client.class) on the top-level Application class but that doesn't seem to do much. The docs indicate @Required might help too ("Requires classes annotated with the given annotations to be available to the application via package scanning"), so I've tried @Required(beans = Client.class) and combinations of that (entities=...) to no avail. Does anyone know of any workarounds or what I might be doing wrong here?
